lib/evm/instruction.ex

```diff
@@ -312,25 +314,42 @@ defmodule EVM.Instruction do
   """
   @spec run_instruction(instruction, EVM.state, MachineState.t, SubState.t, ExecEnv.t) :: {EVM.state, MachineState.t, SubState.t, ExecEnv.t}
   def run_instruction(instruction, state, machine_state, sub_state, exec_env) do
-    # TODO: Make better / break into smaller sections
+    method = instruction_method(instruction)
+    {args, stack} = instruction_args(instruction, state, machine_state, sub_state, exec_env)
+    machine_state = %{machine_state| stack: stack}
+    op_result = apply(EVM.Instruction.Impl, method, args)
+
+    op_result = if is_integer(op_result) do
+      %{stack: Stack.push(stack, op_result)}
+    else
+      op_result
+    end
+
+    op_result |> merge_state(instruction, state, machine_state, sub_state, exec_env)
+  end
+
+  defp instruction_method(instruction) do
```

Comment:
Maybe this should return a `{module, function, args}` which we can then apply as:

`apply(module, function, args ++ instruction_args)`

This will give us more versatility, and more importantly, let us break `impl.ex` into smaller sub-modules.

Reply:
E.g. `@spec instruction_mfa(instruction) :: mfa()`